gadgets/trace_open/test/unit/trace_open_test.go

```diff
@@ -0,0 +1,138 @@
+
+package tests
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"golang.org/x/sys/unix"
+
+	utilstest "github.com/inspektor-gadget/inspektor-gadget/internal/test"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/testing/gadgetapi"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/testing/match"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/testing/utils"
+)
+
+type ExpectedTraceOpenEvent struct {
+	Comm  string `json:"comm"`
+	Pid   uint32 `json:"pid"`
+	Tid   uint32 `json:"tid"`
+	Uid   uint32 `json:"uid"`
+	Gid   uint32 `json:"gid"`
+	Fd    uint32 `json:"fd"`
+	FName string `json:"fname"`
+}
+
+type testDef struct {
+	runnerConfig  *utilstest.RunnerConfig
+	generateEvent func() (int, error)
+	expectedEntry *ExpectedTraceOpenEvent
+	normalize     func(e *ExpectedTraceOpenEvent)
+}
+
+func TestTraceOpenGadget(t *testing.T) {
+	utilstest.RequireRoot(t)
+	testCases := map[string]testDef{
+		"testcase1": {
+			runnerConfig:  &utilstest.RunnerConfig{},
+			generateEvent: generateEvent,
+			expectedEntry: &ExpectedTraceOpenEvent{
+				Comm:  "unit.test",
+				Pid:   utils.NormalizedInt,
+				Tid:   utils.NormalizedInt,
+				Uid:   0,
+				Gid:   0,
+				Fd:    utils.NormalizedInt,
+				FName: "/dev/null",
+			},
+			normalize: func(e *ExpectedTraceOpenEvent) {
+				utils.NormalizeInt(&e.Pid)
+				utils.NormalizeInt(&e.Tid)
+				utils.NormalizeInt(&e.Fd)
+			},
+		},
+		"testcase2": {
+			runnerConfig: &utilstest.RunnerConfig{},
+			generateEvent: func() (int, error) {
+				fd, err := unix.Open("/tmp/foo.test", unix.O_RDONLY|unix.O_CREAT, 0)
+				if err != nil {
+					return 0, fmt.Errorf("opening file: %w", err)
+				}
+				defer unix.Close(fd)
+
+				badfd, err := unix.Open("/tmp/quux.test", unix.O_RDONLY|unix.O_CREAT, 0)
+				if err != nil {
+					return 0, fmt.Errorf("opening file: %w", err)
+				}
+				defer unix.Close(badfd)
+
+				return fd, nil
+			},
+			expectedEntry: &ExpectedTraceOpenEvent{
+				Comm:  "unit.test",
+				Pid:   utils.NormalizedInt,
+				Tid:   utils.NormalizedInt,
+				Uid:   0,
+				Gid:   0,
+				Fd:    utils.NormalizedInt,
+				FName: "/tmp/foo.test",
+			},
+			normalize: func(e *ExpectedTraceOpenEvent) {
+				utils.NormalizeInt(&e.Pid)
+				utils.NormalizeInt(&e.Tid)
+				utils.NormalizeInt(&e.Fd)
+			},
+		},
+	}
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			timeout := 10 * time.Second
+			api, _ := gadgetapi.NewGadget("ghcr.io/inspektor-gadget/gadget/trace_open:latest", timeout)
+			params := map[string]string{
+				"operator.oci.ebpf.uid": "0",
+			}
+
+			api.StartGadget(params)
+			runner := utilstest.NewRunnerWithTest(t, testCase.runnerConfig)
+			utilstest.RunWithRunner(t, runner, func() error {
+				testCase.generateEvent()
+				// Wait for the event to be processed
+				time.Sleep(1 * time.Second)
+				output := api.Output()
+				match.MatchEntries(t, match.JSONSingleArrayMode, string(output), testCase.normalize, testCase.expectedEntry)
```

Comment:
In the issue #3195 I proposed to use the [match](https://github.com/inspektor-gadget/inspektor-gadget/tree/v0.30.0/pkg/testing/match) package to validate events, however I'm not sure it's a good fit for unit tests. As I mentioned above, one big different between unit and integration tests is that in the former we want to be as exhaustive as possible, it means to check fields that change at runtime, like pid, tid, fd, etc. The internal/test/helpers.go package implements functions that are more flexible because they receive some callbacks, so perhaps they're better for this purpose.